src/source/tile.js

```diff
@@ -206,78 +181,58 @@ class Tile {
-        this.cameraToTileDistance = cameraToTileDistance;
-        this.showCollisionBoxes = source.map.showCollisionBoxes;
-        this.placementSource = source;
+    placeLayer(showCollisionBoxes: boolean, collisionIndex: CollisionIndex, layer: any, posMatrix: Float32Array) {
+        const bucket = this.getBucket(layer);
+        const collisionBoxArray = this.collisionBoxArray;
 
-        this.state = 'reloading';
-        this.placementThrottler.invoke();
+        if (bucket && bucket instanceof SymbolBucket && collisionBoxArray) {
+            collisionIndex.setMatrix(posMatrix);
+
+            const pitchWithMap = bucket.layers[0].layout['text-pitch-alignment'] === 'map';
+            const pixelRatio = pixelsToTileUnits(this, 1, collisionIndex.transform.zoom);
+            const labelPlaneMatrix = projection.getLabelPlaneMatrix(posMatrix, pitchWithMap, true, collisionIndex.transform, pixelRatio);
+            PlaceSymbols.place(bucket, collisionIndex, showCollisionBoxes, collisionIndex.transform.zoom, pixelRatio, labelPlaneMatrix, this.coord.id, collisionBoxArray);
+        }
     }
 
-    _immediateRedoPlacement() {
-        this.placementSource.dispatcher.send('redoPlacement', {
-            type: this.placementSource.type,
-            uid: this.uid,
-            source: this.placementSource.id,
-            angle: this.angle,
-            pitch: this.pitch,
-            cameraToCenterDistance: this.cameraToCenterDistance,
-            cameraToTileDistance: this.cameraToTileDistance,
-            showCollisionBoxes: this.showCollisionBoxes
-        }, (_, data) => {
-            if (this.state !== 'reloading') return;
-
-            this.state = 'loaded';
-            this.reloadSymbolData(data, this.placementSource.map.style);
-            this.placementSource.fire('data', {tile: this, coord: this.coord, dataType: 'source'});
-            // HACK this is nescessary to fix https://github.com/mapbox/mapbox-gl-js/issues/2986
-            if (this.placementSource.map) this.placementSource.map.painter.tileExtentVAO.vao = null;
-
-            if (this.redoWhenDone) {
-                this.state = 'reloading';
-                this.redoWhenDone = false;
-                this._immediateRedoPlacement();
+    commitPlacement(collisionIndex: CollisionIndex, collisionFadeTimes: any, angle: number) {
+        // Start all collision animations at the same time
+        for (const id in this.buckets) {
+            const bucket = this.buckets[id];
+            if (bucket instanceof SymbolBucket) {
+                PlaceSymbols.updateOpacities(bucket, collisionFadeTimes);
+                bucket.sortFeatures(angle);
             }
-        }, this.workerID);
+        }
+
+        // Don't update the collision index used for queryRenderedFeatures
+        // until all layers have been updated to the same state
+        if (this.featureIndex) {
+            this.featureIndex.setCollisionIndex(collisionIndex);
```

Comment:
Does this need to be set once for every tile? Or should this be done at the style level?

Reply:
Are you thinking you would pass the appropriate `CollisionIndex` in as a parameter to `FeatureIndex#query`? The chain of calls for `queryRenderedFeatures/Symbols` is somewhat long and confusing, and I feel like it might be kind of hard to keep track of what the "right" collision index is.